Feat loops (How many Stars and complexity analysis of short code snippets)

Fabianstw opened the pull request:
[Test location](https://tcs.uni-frankfurt.de/algo-learn-testing/pull_request_22/en/)

I added more variation for Loops.ts (manly how many stars are printed and an introduction to complexity analysis of short code-snippets) 

LoopsUtilsStars includes:
- more complex if statements 
- 3 different loop types for code snippets
  - simple for loop 
  - double for loop (can be quite difficult)
    - those can be in relation to another 
    - up to two if statement 
    - maybe a continue (only first loop) or a break (second loop) to test user understanding of those two statements
  - while loop 
    - different options on how the condition could be (more details see `export const allWhileCases: string[]` in `loopsUtils.ts` line `37` (including comments with explanation))
  
This PR should resolve issue #9  

Tests for loopsUtilsStars are included. 

Some functions already have functionality for future Big-O runtime questions (currently not included in this PR)

Additional I changed the linter settings, so generator names can match PascalCase

holgerdell commented:
Irgendwas passt hier nicht zusammen:
<img width="649" alt="Screenshot 2024-05-04 at 15 07 31" src="https://github.com/goethe-tcs/algo-learn/assets/467425/0f713edf-457c-438e-b239-834df5945efe">
Die Eingabe t wird nirgends benutzt und D(t) wird definiert, aber nicht benutzt.

Soll das Textfeld mit
"D(t) = [textfeld]"
beschriftet sein, oder sollen t und D(t) komplett gelöscht werden, da in dieser Aufgabenvariante nach einer ganzen Zahl gefragt ist?

Fabianstw commented:
Da stimme ich dir zu! Ich würde die Aufgabenstellung anpassen zu: 
`Consider the following procedure: 
Code.... 
What is the number of stars the procedure prints?` Den Prompt würde ich weglassen. Dann wäre es sehr kurz und simpel. 

Wir sollten dies auf dem branch `stable` auch anpassen, da dort aktuell die gleiche Aufgabenstellung ist.

holgerdell commented:
> Da stimme ich dir zu! Ich würde die Aufgabenstellung anpassen zu: `Consider the following procedure: Code.... What is the number of stars the procedure prints?` Den Prompt würde ich weglassen. Dann wäre es sehr kurz und simpel.

Klingt gut.

AbcAeffchen reviewed (changes requested):
Thanks for the update on the pseudocode. The display of pseudocode using latex looks nice. 

There is some strange behavior when interacting with the code.

- when I select some text and copy (ctrl + c) I get this:
  ```
  1:1:  function f (x):function f (x):
  2:2:    if  x≤1 thenif  x≤1 then
  ```
- when I click the copy button I get this
  ```
    1: \textbf{function}\text{ }\textit{f}\text{ }(x):
    2:     \textbf{if }\text{ }x \leq 1\text{ }\textbf{then}
    3:         \text{print}(\texttt{***})
    4:     \textbf{else}
    5:         \text{print}(\texttt{***})
    6:         \textit{f }(x - 4)
    7:     \text{print}(\texttt{*})
    8: ## ## ## ## ##
    9: {\color{DarkOrchid} \textbf{function}}\text{ }{\color{DarkCyan} \textit{f}}\text{ }({\color{IndianRed}x}):
    10:     {\color{DarkOrchid} \textbf{if }}\text{ }{\color{IndianRed}x} \leq 1\text{ }{\color{DarkOrchid} \textbf{then}}
    11:         {\color{SeaGreen} \text{print}}({\color{Orange} \texttt{***}})
    12:     {\color{DarkOrchid} \textbf{else }}
    13:         {\color{SeaGreen} \text{print}}({\color{Orange} \texttt{***}})
    14:         {\color{DarkCyan} \textit{f }}({\color{IndianRed}x} - 4)
    15:     {\color{SeaGreen} \text{print}}({\color{Orange} \texttt{*}})
  ```

So I'm not sure if the actual latex export should look that way. For an explicit latex copy I probably would prefer an Algpseudocodex-environment.

Currently rendering pseudocode in different styles (like the Algpseudocodex) is not possible since the generator has to provide the different code snippets it seems (like a color and non-color version). This should be moved to the front end and split up into a render function that takes the pseudocode object and a style object. The style object should then contain thinks like:

a pseudocode function object is rendered as `function( .. ){ .. }` with bold letters and color x

-- or --

a pseudocode function object is rendered as `fun( .. ):` with italic letters and color y

I added some more comments in the code. I also only looked at the parts related to the pseudocode and skipped the actual question parts. I would catch this up when the other things are are taken care of. Since the pseudocode stuff is quite a lot, it probably would have been better to give it its own branch/PR.

Fabianstw commented on shared/src/question-generators/example/example.ts:
Yes, I added this file for testing and committed it, so you can have a better example of the new pseudocode highlighting. I'll take it out before we can merge

Fabianstw commented on shared/src/question-generators/time/pseudoCodeUtils.ts:
I understand what you mean. I do already have thought about it, and I don't like the way, that the backend currently has to provide two different versions. 

The problem I have is, that we would need to change the generator type to also provide an PseudoCode object. I would keep some mark inside the question text, and then pass a `PseudoCode Object` down via `useContext`? 

Also I am not sure how we could color everything correctly. Colorising keywords like `while`, `for` and so could be done via regex matching. But some other code statements could be difficult, like `variable` or `strings`. 
The backend directly knows, what a variable or a string is. Maybe the generator should already mark some keywords, following the fronted-end can easily add a color schema. Marking like `\\variable{a}`?

AbcAeffchen replied:
I see the issue. The question needs to be serializable. But the Pseudocode object should be serializable as json string and then the front-end can just get back the object.
I don't know if we want to have a special code-block for this e.g. ` ```pseudocode`?

holgerdell replied:
Yes, ideally this would be a serialization that is itself human- and machine-readable pseudocode. However, it does not really matter and we can change it later, so JSON is fine here. It should use a ```pseudocode block, yes.

holgerdell replied:
Rendering the pseudocode does not require React's context. Try to use context only in the very cases that the object will need to change the state of the UI. This is the case for an input text field. In the future, it is potentially also the case for a textarea, in which code needs to be written, or interactions with a graph within ```graph -- in such cases, changes to the object will have to be sent up to the form where the answer can be submitted, and this requires context (or prop drilling, which I would like to avoid)

Fabianstw replied:
Okay, my idea currently has been to change the `interface QuestionBase` and add a new parameter `pseudoCode?: PseudoCode`. We don't have to serialise the code and only parse it once or when the user changes the color styling. 

But then we would have to use either `context` or `prop drilling`. 
If we use Json we would give up type safety. 

I will program it within ```pseudocode as a json. We can always change it later 👍

AbcAeffchen replied:
Changing the interface would also require to somehow tell where to place the code in the question. And maybe we want to have multiple pseudocode blocks. In the end the questions needs to be serializable because we want to be able to send questions over the net.

The json version makes it easy to serialize and deserialize. An alternative would be to convert it to latex algpseudocodex code. But for this we would need a parser that brings back the pseudocode object and I'm not sure if it is really "human readable".

Fabianstw replied:
I understand the problem.

holgerdell replied:
- First of all: We should do whatever is easiest to maintain and extend.
- Indeed using ```pseudocode + JSON/whatever is giving up type-safety and contradicts with what we said last week in another place. However, Alex and I decided a while ago that all communication between the question generators and the front-end should be objects that are serializable, since they may be sent over the wire in the future. On the other hand, being serializable does not necessarily mean that they need to be serialized and deserialized in the current client-only application. And we are not actually doing this at the moment, the only field that is fully serialized and deserialized is the question markdown text.
- In principle, adding pseudoCodeBlocks?: PseudoCode[] to QuestionBase would be possible, and they are inserted into the appropriate place in Markdown. However, this seems to add a lot of complexity to the QuestionBase type. We will have many different of these special blocks (also think about ```graph, ```tree, etc.), but we want QuestionBase to be generic.

So I see two solutions:
- Fully serialize all special blocks in Markdown: To accept questions sent over the wire, we need to do this anyway.
- Define a generic SpecialBlock type, add a field specialBlocks: SpecialBlock[] to questionBase, and add render components to the front-end for each type of SpecialBlock (which can use `useContext` to affect the state of the surrounding question form).

holgerdell replied:
(..and I don't know what's best. For the moment, markdown ```...``` blocks seem cleanest to me.)

Fabianstw replied:
After programming both options, I believe using pseudocode JSON is the simplest way to present the pseudoCode.

I'll make a few changes and then commit this version, so you can take a look at it.

Fabianstw replied:
How do we want to colorise variables? Should the backend write those into `\\var{x}` for example. When parsing it to correct code (or code we want to display), the frontend can either colorise x or not. 

The `copy` button should work too. When clicking the button LaTeX code from `Algpseudocodex` will be copied and not the code, we create to display on the website

holgerdell replied:
the JSON will probably contain objects of the form { type: "variable", contents: "x" } or similar. The frontend can then choose to colorize it, or not.

Fabianstw replied:
Okay, but then the backend would need to create Objects like:
`const printStatement = ["5+" + { type: "variable", contents: "x" } + "+" + { type: "variable", contents: "y" }]`
When we would mark it it could be:
`const printStatement = "5+\\var{x}+\\var{y}` 
The frontend could match and replace, either with colouring or without.

It would not be as abstract as with objects, but I think it would be simpler.

AbcAeffchen replied:
The object could also look like this: `{var: "x"}`? or you use a helper function like `makeVar("x")` that returns the right object.

Fabianstw replied:
I changed all the types and function correspondingly to your suggestions. The copy button will now save Latex code as `Algpseudocodex` to the clipboard. 
I hope you agree with this :-) 

Inside example.ts is an example for most of the functions. 

I'll add the option, that the front-end can switch between different syntax highlighting schemes (color, bold, italic...). My idea would be to use `useContext` again?, because those values may change and we would not need to drill the styling down over 5-8 functions. 

styling maybe: 
```
const synatxStyle : {
  variable: {
    color: string,
    decoration: "bold" | "italic" | "underline" | "none"
  }, 
  function: {
    ...
  },
  ...
}
```
Colors we could use those: [HTML Color Names](https://www.w3schools.com/tags/ref_colornames.asp)

holgerdell replied:
To keep it simple, I think adding the tailwind className string here would be enough. e.g.
{ variable: "font-bold text-green-500", function: "..."} etc. These should be attached to <span className="...">x</span>

IMO:
- the user should have basically no control over the theme used for syntax highlighting. This just adds unnecessary complexity for us and for the user.
- if however we decide this to be useful anyway, it would be a site-wide setting, probably extending the theme. Theming information (dark/light) is currently provided with the useTheme hook and does not require "context". Context should only be used if the state cannot be global and needs to be attached to a particular component subtree. (-this is the case for questions, because I can imagine the situation that we want to display multiple questions at once)

Fabianstw replied:
ah correct, I forgot about `useTheme `. 

About the tailwind className, I currently write the complete text inside Katex, for that I am using colouring like `{\color{...} ...}` and no `span` attributes. 

If the user should not be able to change the syntax highlighting theme, I would just keep it the way I currently created it. Only change the colours, because (as Alex said) purple is hard to read in dark mode.

Fabianstw commented on front-end/src/components/PseudoCode.tsx:
```diff
@@ -0,0 +1,223 @@
+  )
+}
+
+function createLineNumbers({ index, codeLines }: { index: number; codeLines: number }) {
```
Yes, I change the name

Fabianstw commented on front-end/src/components/PseudoCode.tsx:
```diff
@@ -0,0 +1,223 @@
+            <pre
+              className={`overflow-x-auto whitespace-pre py-3 pl-5 pr-10 font-mono leading-${positionLineHeight} text-gray-900 dark:text-gray-100`}
+            >
+              {(toggleStateColor ? color : code).map((cd, index) => (
```
Changed it 👍

Fabianstw commented on front-end/src/components/PseudoCode.tsx:
```diff
@@ -0,0 +1,223 @@
+function createLineNumbers({ index, codeLines }: { index: number; codeLines: number }) {
+  let codeLine = ""
+  if (codeLines > 0) {
+    codeLine = `${index + 1 >= 10 ? "" : " "}$${index + 1}:$ `
```
Yes, I think this is a simple solution. It only works for 2 digits, but as you say we probably don't want 100 lines of pseudoCode :-) 

change it to `{toggleStateLines && <span className="select-none text-right min-w-10 inline-block">{index + 1}:</span>}` infant of the Markdown.

AbcAeffchen commented on shared/src/question-generators/example/example.ts:
```diff
@@ -14,8 +30,12 @@ import { t, tFunctional, Translations } from "../../utils/translations"
+    text: `Let {{0}} and {{1}} be two - natural numbers. What is the **sum** {{0}}+{{1}}?
+\`\`\`pseudoCode
+{{2}}
+\`\`\`
```
maybe we want to have a function that turns a pseudocode object into such a codeblock with the right name and stringifies the objects. This will be the frequent pattern. And if we want to change it later we only need to change the function.

Fabianstw replied:
Sounds good, I'll add it. But I would still just use `JSON.stringify` to convert the PseudoCode Object into a Json string, right?

AbcAeffchen replied:
yes, but i would do it in that function. So the markdown in the question would just look like `{{1}}` and the code for the data to insert here is `stringifyPseudoCode(pseudoCode)` or something like that.

AbcAeffchen commented on shared/src/question-generators/time/pseudoCodeUtils.ts:
```diff
@@ -0,0 +1,94 @@
+export type PseudoCodeIf = {
+  if: {
+    condition: PseudoCodeString
+    then: PseudoCodeState | PseudoCodeBlock | null
```
`PseudoCodeBlock` already contains `PseudoCodeState` and the `then` part should probably never be empty.

Fabianstw replied:
I added the `null`, so it's more simple to create an if. 

So you can create it straight forward and not reversed. But I can take it out if you want? It's not that important

AbcAeffchen replied:
mh, I see the point. It's probably not that important. But it looks like it is allowed to be empty but after the generator finished some of the properties should not be null. We can also ensure this by reviewing. Maybe @holgerdell has an opinion on this.

AbcAeffchen commented on shared/src/question-generators/time/pseudoCodeUtils.ts:
```diff
@@ -0,0 +1,94 @@
+  if: {
+    condition: PseudoCodeString
+    then: PseudoCodeState | PseudoCodeBlock | null
+    elseif?: Array<{
```
The `then` and the `else` part already allow for another `if` block. Maybe an explicit `elseif` block is not needed? but im not strongly against it.

Fabianstw replied:
The `elseif` block is used to create inline `else if` instead of `else \n__if` and for converting the PseudoCode into Latex, because there is also a difference between `\Elsif` and `Else`.

AbcAeffchen replied:
ok, then leave it as it is.

AbcAeffchen commented on shared/src/question-generators/time/pseudoCodeUtils.ts:
```diff
@@ -0,0 +1,94 @@
+}
+export type PseudoCodeBlock = {
+  block: Array<
+    PseudoCodeState | PseudoCodeBlock | PseudoCodeIf | PseudoCodeWhile | PseudoCodeFor | PseudoCodeForAll
```
Should a `PseudoCodeBlock` allowed to be a array of `PseudoCodeBlock`s? The PseudoCodeBlock does not do anything and is just a container. One can do this in C++ e.g. but this is more for namespace and Constructor/Destructor trickery.

Fabianstw replied:
Yes probably this is not very useful, and we can just add more `PseudoCodeStates` inside a `PseudoCodeBlock`. I'll take it out 👍

AbcAeffchen commented on shared/src/question-generators/time/pseudoCodeUtils.ts:
The pseudocode stuff is independent of the question so it should be placed in `shared/src/utils` and then the `Utils` suffix is not needed.

AbcAeffchen commented on shared/src/utils/parsePseudoCode.ts:
This seems to be part of the frontend as it stringifies a pseudocode object not for reuse but for displaying. So `front-end/src/lib` or `front-end/src/utils` would be a better place.

AbcAeffchen commented on shared/src/utils/parsePseudoCode.ts:
```diff
@@ -0,0 +1,555 @@
+  return typeof pseudoString === "object" && "printString" in pseudoString
+}
+
+function leadingWhitespace(indent: number): string {
```
for the latex part it would probably be better to use something like `\hspace{}` instead of whitespaces. Latex has the option to shrink the size of whitespace if the line is too full.

Fabianstw replied:
I would suggested I take this out from the Latex part, because the package `Algpseudocodex` does indentation  automatically.

AbcAeffchen commented on shared/src/utils/parsePseudoCode.ts:
A general suggestion: The generates three string representations of the pseudocode object. During generation the string is also split up at the `\n` characters and joined again and the receiving react function splits up into lines again.
Maybe it is easier to generate arrays of codelines. This save some string operations and all the `+ "\n"`.

Fabianstw replied:
Yes correct, I'll change this. 

I separated it by `\n` first because I displayed the string inside: \`\`\`pseudoCode, but now we parse it into a string later I can just create it as an array<string>

AbcAeffchen commented on shared/src/utils/parsePseudoCode.ts:
```diff
@@ -0,0 +1,555 @@
+  keyWordsColor,
+  printStatementColor,
+  variableColor,
+} from "@/components/DrawPseudoCode.tsx"
```
In the long run we should think about where to put the colors. There should be a folder with style definitions.

Fabianstw commented:
I implemented your feedback, feel free to take a look again :-) 

Also: 
- Changed the color for keyWords to something blue, I think it's better visible in dark and light mode. 
- Added `\\[0.1em]` to each Katex line, because two fractions above each other has been very tight.  (Could also use `mt-0.5`)

I thought I reset `example.ts` and `questionsSelection.ts`, but somehow it's still using the old format, I don't know why ... 
Do you have an idea?

holgerdell commented on front-end/src/components/DrawPseudoCode.tsx:
```diff
@@ -0,0 +1,197 @@
+            <pre
+              className={`overflow-x-auto whitespace-pre py-3 pl-5 pr-10 font-mono leading-normal text-gray-900 dark:text-gray-100 ${toggleStateColor ? "hidden" : ""}`}
+            >
+              {createCodeJSX(pseudoCodeString, isTextVisible, toggleStateLines, numCodeLines)}
```
sollte besser eine React-Komponente sein.

holgerdell commented:
- with weglassen.
- nicht zwischen def proc func etc wechseln.
- graue Zeilennummern, fixe breite
- Tooltip einheitlich
- Mit/ohne Farben memoisieren: https://react.dev/reference/react/useMemo#skipping-re-rendering-of-components
- maxheight und isTextVisible wahrscheinlich als state weglassen

AbcAeffchen commented on front-end/src/utils/parsePseudoCode.ts:
This file seems to render code objects, not parse. The name is misleading. 

It also does most of the things multiple times (normal, color, latex). 
An alternative would be to have a settings object that has all necessary information for a keyword and then tell the function which version is needed. Then one can call the function 3 times to cache the result.

Fabianstw replied:
Name is indeed misleading, renamed it to `renderPseudoCode.ts`. 

Yes the Strings are created 3 times, but they are all a bit different. It think it's simpler to create all three the same time (without a settings object), instead of creating 3 `if-statements` to differ between the different setting options. 

But if we have more setting options in the future I agree that I would be better to call the renderPseudoCode `n` times to create `n` different results. 

For now I would add this as a todo?

Fabianstw commented on front-end/src/components/DrawPseudoCode.tsx:
```diff
@@ -0,0 +1,195 @@
+ *
+ * @param code
+ */
+function adding015EM(code: string): string {
```
Right, changed the name to `addBiggerLneSpacing`

Fabianstw commented:
> * with weglassen.
Habe dies durch Interaktionen durch Mengen ersetzt

> * nicht zwischen def proc func etc wechseln.
Ist nur noch `function`

> * graue Zeilennummern, fixe breite
👍 

> * Tooltip einheitlich
habe deins aus main gemergt 

> * Mit/ohne Farben memoisieren: https://react.dev/reference/react/useMemo#skipping-re-rendering-of-components
Habe hier nicht memoriert habe nur ein paar andere "überflüssige" Punkte weggelassen. Es läuft jetzt deutlich flüssiger 

> * maxheight und isTextVisible wahrscheinlich als state weglassen
Ist raus 👍

Fabianstw commented:
Ich muss im `Loops` Ordner noch den Code sauberer schreiben. Aber über den PseudoCode Teil könnt ihr gerne drüber gucken :-)

holgerdell commented:
Das gefällt mir jetzt ziemlich gut! Nur einen Änderungswunsch habe ich jetzt noch: Bitte auf sprachspezifische Schlüsselwörter verzichten (`break` / `continue`).

Ansonsten kannst du es erstmal mergen, damit der Code-Generator schonmal drin ist.
